[TIMOB-17467] Android: Add setRequestHeaders() to WebView

m1ga opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-17467

**Optional Description:**

```
var win = Ti.UI.createWindow();
var webView = Ti.UI.createWebView({
    requestHeaders: {
        "User-Foo": "FooBar",
        "foo": "bar",
        "other": "string"
    },
    url: "https://httpbin.org/headers"
});
var btn = Ti.UI.createButton({
    title: "Reload",
    bottom: 50
});

btn.addEventListener("click", function() {
    webView.reload();
});

var btnChange = Ti.UI.createButton({
    title: "Change Header",
    bottom: 0
});

btnChange.addEventListener("click", function() {
    var date = new Date().getTime();
    webView.setRequestHeaders({
        "User-Foo": "FooBar_" + date,
        "foo": "bar_" + date,
        "other": "string_" + date
    });
});

win.add(webView);
win.add(btnChange);
win.add(btn);
win.open();
```

m1ga committed:
android: add setRequestHeaders({});

build commented:
Can one of the admins verify this patch?

abada commented:
Very nice 👍

abada commented:
What about iOS ?

m1ga committed:
docs

hansemannn commented on apidoc/Titanium/UI/WebView.yml:
```diff
@@ -272,6 +272,17 @@ methods:
         summary: Basic auth password.
         type: String
+        
+  - name: setRequestHeaders
```
should be a property and also return a getter to receive already stored request-headers. Otherwise, I can already see the improvement-ticket coming 😉

m1ga committed:
add getter

m1ga committed:
change docu

m1ga commented:
added `getter` and changed the documentation!

```
$.index.addEventListener("open", function(e) {
    $.www.setRequestHeaders({
        "User-Blub": "blablabla",
        "something": "string1",
        "other": "string2"
    });
    $.www.setUrl("http://localhost/test.php");
    
    console.log("Len: " + Object.keys($.www.getRequestHeaders()).length);   
    console.log($.www.getRequestHeaders());
    console.log( $.www.requestHeaders );
});

$.index.open();
```

will return:
```
[INFO]  Len: 3
[INFO]  {"User-Blub":"blablabla","something":"string1","other":"string2"}
[INFO]  {"User-Blub":"blablabla","something":"string1","other":"string2"}
```

hansemannn commented:
Sweet! Will plan the iOS-part then. @ashcoding to review for 6.1.0 :-)

m1ga commented:
@hansemannn if you want to change something let me know and I'll adapt it.

ashcoding commented:
@m1ga is there any alternative website that I can use to easily test this? Perhaps something similar to httpbin.org?

m1ga commented:
@ashcoding never heard about httpbin before. Looks awesome! https://httpbin.org/headers should work. Or try my test-page at http://migaweb.de/test.php

ashcoding commented:
I'm trying to do this in a native app:
```
var webView = Ti.UI.createWebView({
    url: "https://httpbin.org/headers",
});

webView.setRequestHeaders({
    "User-Blub": "blablabla",
    "something": "string1",
    "other": "string2"
});
 
var win = Ti.UI.createWindow();
var btn = Ti.UI.createButton({
    title : "Reload"
});
 
btn.addEventListener("click", function() {
    webView.reload();
});
 
win.add(webView);
win.add(btn);
win.open();
```
@m1ga
If I don't add the request headers, it works fine. With the headers, it's not working.
I'm using an Android 6.0, Nexus 6.

I just get a blank screen with nothing loading. Seems like something is going on?

@hansemannn could you try this whenever you have the time as well?

m1ga commented:
@ashcoding if the setRequestHeaders is inside the button click event it works fine. Will check that tomorrow and fix the bug!

hansemannn commented:
I made that test to explicitly validate that the headers persist after reloading the page. So we just need to ensure iOS and Android behave the same.

m1ga commented:
@ashcoding the problem is occuring with `setUserAgent()` too and is mentioned here: https://jira.appcelerator.org/browse/TIMOB-17139
with a workaround and wontfix :cry: . But the workaround works for this PR, too. But I'm not a big fan of this workaround or it should be mentioned in the documentation. 

It would also be good to add a `requestHeaders` parameter you can add to the create method so that in your example you would add it right away and not call the method after creation.

m1ga committed:
adding parameter

m1ga committed:
update documentation

m1ga commented:
So this demo works fine now:
```
var win = Ti.UI.createWindow();
var webView = Ti.UI.createWebView({
    requestHeaders: {
        "User-Foo": "FooBar",
        "foo": "bar",
        "other": "string"
    },
    url: "https://httpbin.org/headers"
});
var btn = Ti.UI.createButton({
    title: "Reload",
    bottom: 50
});

btn.addEventListener("click", function() {
    webView.reload();
});

var btnChange = Ti.UI.createButton({
    title: "Change Header",
    bottom: 0
});

btnChange.addEventListener("click", function() {
    var date = new Date().getTime();
    webView.setRequestHeaders({
        "User-Foo": "FooBar_" + date,
        "foo": "bar_" + date,
        "other": "string_" + date
    });
});

win.add(webView);
win.add(btnChange);
win.add(btn);
win.open();
```

* requestHeaders property are set during creation
* custom headers survive a reload (press button `reload`, output stays the same)
* header can be changed with `change header` + `reload` to display the output. _Just pressing `change header` won't automatically reload the page_

I've added some remarks about how to use setRequestHeaders() method on Android and added the iOS platform so you can merge the iOS PR afterwards

ashcoding commented:
Will you be addressing the issue that I brought up?
Alternatively, since you have added this info in the docs, perhaps this is okay.

@hansemannn what do you think?

hansemannn commented:
I'm fine with merging it, since - as he indicated - it is a general problem with Android-webviews that should be addressed separately.

ashcoding commented:
Roger. I'll give this PR one more review and get it merged if everything is good.

ashcoding commented:
Code functionally tested and reviewed.
LGTM.
**Approved**

ashcoding commented:
Thank you @m1ga
